File: serene/src/Serene.Web/Modules/ServerTypes/Administration/UserPermissionRow.ts
import { fieldsProxy } from "@serenity-is/corelib";

export interface UserPermissionRow {
    UserPermissionId?: number;
    UserId?: number;
    PermissionKey?: string;
    Granted?: boolean;
    Username?: string;
    User?: string;
}

export abstract class UserPermissionRow {
    static readonly idProperty = 'UserPermissionId';
    static readonly nameProperty = 'PermissionKey';
    static readonly localTextPrefix = 'Administration.UserPermission';
    static readonly deletePermission = 'Administration:Security';
    static readonly insertPermission = 'Administration:Security';
    static readonly readPermission = 'Administration:Security';
    static readonly updatePermission = 'Administration:Security';

    static readonly Fields = fieldsProxy<UserPermissionRow>();
}

File: serene/src/Serene.Web/Modules/Administration/UserPermission/UserPermissionRow.cs
namespace Serene.Administration;

[ConnectionKey("Default"), Module("Administration"), TableName("UserPermissions")]
[DisplayName("UserPermissions"), InstanceName("UserPermissions")]
[ReadPermission(PermissionKeys.Security)]
[ModifyPermission(PermissionKeys.Security)]
public sealed class UserPermissionRow : Row<UserPermissionRow.RowFields>, IIdRow, INameRow, IUserPermissionRow
{
    [DisplayName("User Permission Id"), Identity, IdProperty]
    public long? UserPermissionId { get => fields.UserPermissionId[this]; set => fields.UserPermissionId[this] = value; }

    [DisplayName("User Id"), NotNull, ForeignKey("Users", "UserId"), LeftJoin("jUser")]
    public int? UserId { get => fields.UserId[this]; set => fields.UserId[this] = value; }

    [DisplayName("Permission Key"), Size(100), NotNull, QuickSearch, NameProperty]
    public string PermissionKey { get => fields.PermissionKey[this]; set => fields.PermissionKey[this] = value; }

    [DisplayName("Grant")]
    public bool? Granted { get => fields.Granted[this]; set => fields.Granted[this] = value; }

    [DisplayName("User Username"), Expression("jUser.[Username]")]
    public string Username { get => fields.Username[this]; set => fields.Username[this] = value; }

    [DisplayName("User Display Name"), Expression("jUser.[DisplayName]")]
    public string User { get => fields.User[this]; set => fields.User[this] = value; }

    Field IUserPermissionRow.UserIdField => fields.UserId;
    StringField IUserPermissionRow.PermissionKeyField => fields.PermissionKey;
    BooleanField IUserPermissionRow.GrantedField => fields.Granted;

    public class RowFields : RowFieldsBase
    {
        public Int64Field UserPermissionId;
        public Int32Field UserId;
        public StringField PermissionKey;
        public BooleanField Granted;

        public StringField Username;
        public StringField User;
    }
}